platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/branching/DefaultBranchDetector.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.event.branching;
+
+import com.swirlds.common.platform.NodeId;
+import com.swirlds.platform.consensus.EventWindow;
+import com.swirlds.platform.event.GossipEvent;
+import com.swirlds.platform.system.address.AddressBook;
+import com.swirlds.platform.system.events.EventDescriptor;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A standard implementation of {@link BranchDetector}.
+ */
+public class DefaultBranchDetector implements BranchDetector {
+
+    /**
+     * The current event window.
+     */
+    private EventWindow currentEventWindow;
+
+    /**
+     * The node IDs of the nodes in the network in sorted order, provides deterministic iteration order.
+     */
+    private final List<NodeId> nodes = new ArrayList<>();
+
+    /**
+     * The most recent non-ancient events for each node (not present or null if there are none).
+     */
+    private final Map<NodeId, EventDescriptor> mostRecentEvents = new HashMap<>();
+
+    /**
+     * Create a new branch detector.
+     *
+     * @param currentRoster the current roster
+     */
+    public DefaultBranchDetector(@NonNull final AddressBook currentRoster) {
+        nodes.addAll(currentRoster.getNodeIdSet());
+        Collections.sort(nodes);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public GossipEvent checkForBranches(@NonNull final GossipEvent event) {
+        if (currentEventWindow == null) {
+            throw new IllegalStateException("Event window must be set before adding events");
+        }
+
+        if (currentEventWindow.isAncient(event)) {
+            // Ignore ancient events.
+            return null;
+        }
+
+        final NodeId creator = event.getCreatorId();
+        final EventDescriptor previousEvent = mostRecentEvents.get(creator);
+        final EventDescriptor selfParent = event.getHashedData().getSelfParent();
+
+        final boolean branching = !(previousEvent == null || previousEvent.equals(selfParent));
```

Comment:
When we restart from round X, we are guaranteed to have access to all judges for round X and all of their non-ancient ancestors. This class can be missing events because they are ancient, but restart/reconnect doesn't change the calculus on that.

That being said, this IS a best effort algorithm when applied to preconsensus events as it currently is. That's because not all nodes may have an event before it reaches consensus, so it's plausible that I may detect a branch but nobody else ever will.

In order to make this more than best effort, we will have to plug this into the stream of consensus events. Something we will do eventually, but not in the short term. This class just generates metrics/logging, so it's ok if there is some wiggle room in the algorithm.